zookeeper-server/src/main/java/org/apache/zookeeper/server/DataTree.java

```diff
@@ -808,74 +840,163 @@ public int compare(PathWithStat left, PathWithStat right) {
      * Produces a paginated list of the children of a given path
      * @param path path of node node to list
      * @param stat stat of the node to list
-     * @param watcher an optional watcher to attach to the node. The watcher is added only once when reaching the end of pagination
-     * @param maxReturned maximum number of children to return. Return one more than this number to indicate truncation
+     * @param watcher an optional watcher to attach to the node. The watcher is added only once
+     *                when reaching the end of pagination
+     * @param maxReturned maximum number of children to return.
      * @param minCzxId only return children whose creation zxid equal or greater than minCzxId
      * @param czxIdOffset how many children with zxid == minCzxId to skip (as returned in previous pages)
-     * @return A list of path with stats
+     * @param nextPage info to be used for the next page call
+     * @return A list of child names of the given path
      * @throws NoNodeException if the path does not exist
      */
-    public List<PathWithStat> getPaginatedChildren(String path, Stat stat, Watcher watcher, int maxReturned,
-                                                   long minCzxId, long czxIdOffset)
+    public List<String> getPaginatedChildren(String path, Stat stat, Watcher watcher, int maxReturned,
+                                             long minCzxId, int czxIdOffset, PaginationNextPage nextPage)
             throws NoNodeException {
         DataNode n = nodes.get(path);
         if (n == null) {
             throw new KeeperException.NoNodeException();
         }
-        synchronized (n) {
-            if (stat != null) {
-                n.copyStat(stat);
-            }
-            PriorityQueue<PathWithStat> childrenQueue;
-            Set<String> actualChildren = n.getChildren();
-            if (actualChildren == null) {
-                childrenQueue = new PriorityQueue<PathWithStat>(1);
-            } else {
-                childrenQueue = new PriorityQueue<PathWithStat>(maxReturned + 1, staticNodeCreationComparator);
-                for (String child : actualChildren) {
-                    DataNode childNode = nodes.get(path + "/" + child);
-                    if (null != childNode) {
-                        final long czxId = childNode.stat.getCzxid();
-
-                        if (czxId < minCzxId) {
-                            // Filter out nodes that are below minCzxId
-                            continue;
-                        }
-
-                        Stat childStat = new Stat();
-                        childNode.copyStat(childStat);
-
-                        // Cannot discard before having sorted and removed offset
-                        childrenQueue.add(new PathWithStat(child, childStat));
+
+        if (maxReturned == Integer.MAX_VALUE && minCzxId <= 0) {
```

Comment:
The only reason we need to separate getting all children and getting partial children is to ensure "the order is the same as before". But this is not necessary. Since the order is not guaranteed for the getChildren call anyway.

I think splitting the logic makes the code looks ugly and hard to reason with. I strongly suggest merging the logic. We can always rely on buildChildrenPathWithStat to get all children if no parameter is configured to limit the return. Then the additional check can be applied to the list for generating a paginated result.

Note that if we split the logic, then we may have 2 tries on the server side. One for trying to fetch everything, another one for trying to fetch the first page (if the first try fails). This may be a performance concern.

Reply:
The most important reason I'd like to separate it is performance/behavior, not the order.
99% of the case the children can be returned in a page and so won't go to the 2nd part.

The 2nd part always sort. And note that sorting is in the synchronized block. Imagine if the node has many children eg. 30K, sorting would cost more time than directly returning the list. And this cost would impact other synchronized threads and read/write performance.

"then we may have 2 tries on the server side": Only 1% of the case would just go to both 1st and 2nd parts. I don't think this is a performance concern, compared to every time needs sorting.

So rather than some more code, I would still prefer to ensure the performance is optimal.